### PR TITLE
Mainline development mode

### PR DESCRIPTION
**Pull request by JakeGinnivan**
This builds on top of #832

It's pretty rough but will ship this as part of 4.0.0 beta if it looks ok

### CODE REVIEW

**Comment by pascalberger** [comment]
The branching strategy images still contain `-unstable` tags. Since `alpha` is the default in this PR I would change this in the images too.

File: src/GitVersionCore/AssemblyInfo.cs [hunk 1]
```diff
@@ -1,11 +1,13 @@
-﻿using System.Reflection;
+using System.Reflection;
 using System.Runtime.CompilerServices;
 
 [assembly: AssemblyTitle("GitVersionCore")]
 [assembly: AssemblyProduct("GitVersion")]
-[assembly: AssemblyVersion("1.0.0")]
-[assembly: AssemblyFileVersion("1.0.0")]
+[assembly: AssemblyVersion("4.0.0.0")]
```
Comment (pascalberger) [inline src/GitVersionCore/AssemblyInfo.cs]: Isn't this set through GitVersion in the build process and should be kept to 1.0.0 in the repo?

File: src/GitVersionCore/AssemblyInfo.cs [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
+[assembly: AssemblyFileVersion("4.0.0.0")]
 [assembly: InternalsVisibleTo("GitVersionTask.Tests")]
 [assembly: InternalsVisibleTo("GitVersion")]
 [assembly: InternalsVisibleTo("GitVersionCore.Tests")]
 [assembly: InternalsVisibleTo("GitVersionExe.Tests")]
+
+[assembly: AssemblyInformationalVersion("4.0.0-MainlineDevelopmentMode.1+972.Branch.MainlineDevelopmentMode.Sha.be15618ebc310451a745278841148c9f1fcc7108")]
```
Comment (pascalberger) [inline src/GitVersionCore/AssemblyInfo.cs]: Ditto (Should be set in the build process)

File: src/GitVersionExe/AssemblyInfo.cs [hunk 3]
```diff
@@ -1,10 +1,12 @@
-﻿using System.Reflection;
+using System.Reflection;
 using System.Runtime.CompilerServices;
 
 [assembly: AssemblyTitle("GitVersion")]
 [assembly: AssemblyProduct("GitVersion")]
-[assembly: AssemblyVersion("1.0.0")]
-[assembly: AssemblyFileVersion("1.0.0")]
+[assembly: AssemblyVersion("4.0.0.0")]
+[assembly: AssemblyFileVersion("4.0.0.0")]
```
Comment (pascalberger) [inline src/GitVersionExe/AssemblyInfo.cs]: Ditto (Should be set in the build process)

File: src/GitVersionExe/AssemblyInfo.cs [hunk 4]
```diff
@@ -1,10 +1,12 @@
...
 [assembly: InternalsVisibleTo("GitVersionTask.Tests")]
 [assembly: InternalsVisibleTo("AcceptanceTests")]
 [assembly: InternalsVisibleTo("GitVersionExe.Tests")]
+
+[assembly: AssemblyInformationalVersion("4.0.0-MainlineDevelopmentMode.1+972.Branch.MainlineDevelopmentMode.Sha.be15618ebc310451a745278841148c9f1fcc7108")]
```
Comment (pascalberger) [inline src/GitVersionExe/AssemblyInfo.cs]: Ditto (Should be set in the build process)

File: src/GitVersionTask/AssemblyInfo.cs [hunk 5]
```diff
@@ -1,6 +1,7 @@
...
+using System.Reflection;
 
 [assembly: AssemblyTitle("GitVersionTask")]
 [assembly: AssemblyProduct("GitVersionTask")]
-[assembly: AssemblyVersion("1.0.0")]
-[assembly: AssemblyFileVersion("1.0.0")]
\ No newline at end of file
+[assembly: AssemblyVersion("4.0.0.0")]
+[assembly: AssemblyFileVersion("4.0.0.0")]
+[assembly: AssemblyInformationalVersion("4.0.0-MainlineDevelopmentMode.1+972.Branch.MainlineDevelopmentMode.Sha.be15618ebc310451a745278841148c9f1fcc7108")]
```
Comment (pascalberger) [inline src/GitVersionTask/AssemblyInfo.cs]: Ditto (Should be set in the build process)

**Comment by pascalberger** [comment]
We should also make sure to document the new mainline mode properly. There is already a [mainline documentation](https://github.com/GitTools/GitVersion/blob/master/docs/reference/mainline-development.md) available which should be updated with these changes.

File: src/GitVersionCore/AssemblyInfo.cs (hunk 1 above)
Comment (JakeGinnivan) [inline src/GitVersionCore/AssemblyInfo.cs]: Ran build script locally, forgot to reset these changes

File: docs/versioning-mode.md [hunk 6]
```diff
@@ -1,2 +1,40 @@
...
+
+## Continuous deployment
+Sometimes you just want the version to keep changing and continuously deploy. A good case for this is when using Octopus deploy, as you cannot publish the same version of a package into the same feed.
+
+For this mode we followed the logic in this blog post by Xavier Decoster on the issues of incrementing automatically - http://www.xavierdecoster.com/semantic-versioning-auto-incremented-nuget-package-versions
+
+As such we force a pre-release tag on all branches, this is fine for applications but can cause problems for libraries. As such this mode may or may not work for you, which leads us into a new mode in v4. Mainline development.
+
+## Mainline development
+Mainline development works more like the Continuous Delivery mode, except that it tells GitVersion to *infer* releases from from event.
```
Comment (stocksr) [inline docs/versioning-mode.md]: typo: from from

**Comment by aarondandy** [comment]
I did a local build and tried it out, worked great and as expected during my quick testing.

My :sweet_potato:
```
assembly-versioning-scheme: MajorMinorPatch
mode: Mainline
branches:
  master:
    tag: ''
ignore:
  sha: []
```

I did get two test failures though when running build.ps1:
Test Count: 451, Passed: 449, Failed: 2, Inconclusive: 0, Skipped: 0
```

**Pull request merged**